building_sim_plugins/building_ignition_plugins/src/slotcar.cpp

```diff
@@ -201,9 +239,53 @@ std::vector<Eigen::Vector3d> SlotcarPlugin::get_obstacle_positions(
   return obstacle_positions;
 }
 
+void SlotcarPlugin::charge_state_cb(const ignition::msgs::Selection& msg)
+{
+  dataPtr->charge_state_cb(msg.name(), msg.selected());
+}
+
+// First element of msg should be the slotcar, and the second should be the payload
+void SlotcarPlugin::item_dispensed_cb(const ignition::msgs::UInt64_V& msg)
+{
+  if (msg.data_size() == 2 && msg.data(0) == _entity)
+  {
+    Entity new_payload = msg.data(1);
+    this->_payloads.insert(new_payload);
+  }
+}
+
+void SlotcarPlugin::item_ingested_cb(const ignition::msgs::Entity& msg)
+{
+  if (msg.IsInitialized() && _payloads.find(msg.id()) != _payloads.end())
+  {
+    _payloads.erase(msg.id());
+  }
+}
+
 void SlotcarPlugin::PreUpdate(const UpdateInfo& info,
   EntityComponentManager& ecm)
 {
+  // Read from components that may not have been initialized in configure()
+  if (first_iteration)
+  {
+    Entity parent = _entity;
+    while (ecm.ParentEntity(parent))
+    {
+      parent = ecm.ParentEntity(parent);
+    }
+    if (ecm.EntityHasComponentType(parent,
+      components::PhysicsEnginePlugin().TypeId()))
+    {
+      std::string physics_plugin_name =
+        ecm.Component<components::PhysicsEnginePlugin>(parent)->Data();
+      if (plugin_names.count(physics_plugin_name))
+      {
+        phys_plugin = plugin_names[physics_plugin_name];
+      }
```

Comment:
Same as above, can be replaced to do a single lookup through an iterator. Also I'd generally recommend against using the `operator[]` since it will do insertions when the key is not found (even though it's not an issue here). If you went for the iterator approach you should also be able to make the variable `const` I believe.